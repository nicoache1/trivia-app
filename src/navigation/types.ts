import { ParamListBase } from '@react-navigation/native'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React from 'react'

import { Routes } from './routes'

export interface ParamList extends ParamListBase {
  [Routes.Welcome]: undefined
  [Routes.ConfigureGame]: undefined
  [Routes.Game]: undefined
  [Routes.Results]: undefined
  [Routes.AnimatedSplash]: undefined
}

export type SceneProps<T extends keyof ParamList> = React.FC<
  NativeStackScreenProps<ParamList, T>
>

export type NavigationProp<T extends keyof ParamList> = NativeStackScreenProps<
  ParamList,
  T
>['navigation']
