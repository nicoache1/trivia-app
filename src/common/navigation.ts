import {
  NavigationContainerRef,
  ParamListBase,
  StackActions,
} from '@react-navigation/native'
import * as React from 'react'

import { ParamList } from '../navigation/types'

export const navigationRef: React.RefObject<NavigationContainerRef<ParamList>> =
  React.createRef()

export function navigate(name: string, params?: ParamListBase[string]) {
  navigationRef.current?.navigate(name, params)
}

export function push(name: string, params?: ParamListBase[string]) {
  navigationRef.current?.dispatch(StackActions.push(name, params))
}

export function popToTop() {
  navigationRef.current?.dispatch(StackActions.popToTop())
}

export function pop() {
  navigationRef.current?.dispatch(StackActions.pop())
}
