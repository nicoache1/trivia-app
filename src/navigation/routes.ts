export enum Routes {
  Game = 'Game',
  ConfigureGame = 'ConfigureGame',
  AnimatedSplash = 'AnimatedSplash',
  Results = 'Results',
  Welcome = 'Welcome',
}
