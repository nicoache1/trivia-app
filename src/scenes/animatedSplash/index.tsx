import SplashImage from 'assets/splashShape.png'
import * as SplashScreen from 'expo-splash-screen'
import React, { useLayoutEffect } from 'react'
import { Dimensions, StyleSheet, View } from 'react-native'
import Animated, {
  interpolate,
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'
import { Routes } from 'src/navigation/routes'
import { SceneProps } from 'src/navigation/types'
import { useTheme } from 'src/styles/Theme'

import { styles } from './styles'

const { width, height } = Dimensions.get('window')

export const AnimatedSplash: SceneProps<Routes.AnimatedSplash> = ({
  navigation,
}) => {
  useLayoutEffect(() => {
    navigation.setOptions({ headerShown: false })
  }, [navigation])

  const {
    Theme: { colors },
  } = useTheme()

  const progress = useSharedValue(0)
  const imageAnimation = useSharedValue(0)

  const onImageLoaded = async () => {
    await SplashScreen.hideAsync()
    progress.value = withTiming(1, { duration: 700 }, () => {
      imageAnimation.value = withTiming(1, {}, finished => {
        if (finished) {
          runOnJS(navigation.replace)(Routes.Welcome)
        }
      })
    })
  }

  const animatedStyle = useAnimatedStyle(() => ({
    backgroundColor: colors.SECONDARY,
    transform: [
      { translateY: width / 2 },
      { rotate: `${interpolate(progress.value, [0, 1], [-25, 0])}deg` },
      { translateX: interpolate(progress.value, [0, 1], [-height, 0]) },
      { translateY: -width / 2 },
    ],
    zIndex: 1,
  }))

  const animatedImage = useAnimatedStyle(() => ({
    height: interpolate(imageAnimation.value, [0, 0.5, 1], [70, 80, 0]),
    opacity: interpolate(imageAnimation.value, [0, 1], [1, 0]),
    zIndex: 100,
  }))

  return (
    <View style={styles.container}>
      <Animated.Image
        onLoadEnd={onImageLoaded}
        source={SplashImage}
        style={animatedImage}
        resizeMode={'contain'}
      />
      <Animated.View
        style={[animatedStyle, { ...StyleSheet.absoluteFillObject }]}
      />
    </View>
  )
}
