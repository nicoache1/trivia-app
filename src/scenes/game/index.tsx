import CheckImage from 'assets/check.png'
import CrossImage from 'assets/cross.png'
import React, { useEffect, useLayoutEffect, useRef, useState } from 'react'
import { View } from 'react-native'
import { useSelector } from 'react-redux'
import { Background } from 'src/components/Background'
import { SceneContainer } from 'src/components/SceneContainer'
import { StyledContainer } from 'src/components/StyledContainer'
import { StyledText } from 'src/components/StyledText'
import { Routes } from 'src/navigation/routes'
import { SceneProps } from 'src/navigation/types'
import { useAppDispatch } from 'src/store'
import { useLazyQuestionsQuery } from 'src/store/APIs/questions'
import { setGameResultAction } from 'src/store/storage/game'
import { selectGameConfiguration } from 'src/store/storage/game/selectors'
import { useTheme } from 'src/styles/Theme'
import { AppTypography, ColorPalette } from 'src/styles/types'
import { Question } from 'src/types/questions'
import { v4 as uuidv4 } from 'uuid'

import { Card } from './Card'
import { FinishGame } from './FinishGame'
import { GameButton } from './GameButton'
import { styles } from './styles'

export const Game: SceneProps<Routes.Game> = ({ navigation }) => {
  const game = useSelector(selectGameConfiguration)
  const {
    Theme: { colors },
  } = useTheme()
  const dispatch = useAppDispatch()
  const [triggerQuery, { isLoading, isFetching }] = useLazyQuestionsQuery()

  const ref = useRef<{
    triggerAnimation: (option: boolean) => void
  }>(null)

  const [cards, setCards] = useState<Question[]>([])

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    })
  }, [navigation])

  useEffect(() => {
    async function getQuestions() {
      const data = await triggerQuery({
        amount: game.questionAmount,
        difficulty: game.difficulty,
      }).unwrap()
      setCards(data)
    }

    getQuestions()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const onFinish = (option: boolean) => {
    const question = cards[cards.length - 1]

    dispatch(
      setGameResultAction({
        results: [
          {
            correct: question.correctAnswer === option,
            id: uuidv4(),
            question,
          },
        ],
      }),
    )

    setCards(prevState => {
      return prevState.slice(0, prevState.length - 1)
    })
  }

  const onPressButton = (option: boolean) => () => {
    ref.current?.triggerAnimation(option)
  }

  const finishGame = cards.length === 0 && !isLoading && !isFetching

  const finishCallback = () => navigation.navigate(Routes.Results)

  return (
    <>
      {finishGame && <FinishGame callback={finishCallback} />}
      <SceneContainer
        barStyle={'light-content'}
        style={[styles.container, { backgroundColor: colors.SECONDARY }]}>
        <Background />
        <StyledContainer style={styles.playerContainer}>
          <StyledText color={ColorPalette.WHITE} typography={AppTypography.H1}>
            {game.player}
          </StyledText>
        </StyledContainer>
        <View style={styles.cardContainer}>
          {cards.map((item, index) => (
            <Card
              ref={index === cards.length - 1 ? ref : undefined}
              key={item.id}
              item={item}
              index={index}
              onFinish={onFinish}
            />
          ))}
        </View>
        <StyledContainer style={styles.answerOptionContainer}>
          <GameButton imageSource={CheckImage} onPress={onPressButton(true)} />
          <GameButton imageSource={CrossImage} onPress={onPressButton(false)} />
        </StyledContainer>
      </SceneContainer>
    </>
  )
}
