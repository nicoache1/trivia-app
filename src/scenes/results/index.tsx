import { localization } from 'localization'
import React, { useLayoutEffect } from 'react'
import { Dimensions, FlatListProps, StyleSheet, View } from 'react-native'
import { FlatList } from 'react-native-gesture-handler'
import { Path, Svg } from 'react-native-svg'
import { useSelector } from 'react-redux'

import { SceneContainer } from '../../components/SceneContainer'
import { StyledButton } from '../../components/StyledButton'
import { ButtonMode } from '../../components/StyledButton/ButtonMode'
import { StyledContainer } from '../../components/StyledContainer'
import { Routes } from '../../navigation/routes'
import { SceneProps } from '../../navigation/types'
import { useAppDispatch } from '../../store'
import { restartGameAction } from '../../store/storage/game'
import { selectGameConfiguration } from '../../store/storage/game/selectors'
import { useTheme } from '../../styles/Theme'
import { QuestionResult } from '../../types/questions'
import { Header } from './Header'
import { ResultItem } from './ResultItem'
import { ITEM_HEIGHT } from './ResultItem/styles'
import { styles } from './styles'

const { width, height } = Dimensions.get('window')

const renderItem: FlatListProps<QuestionResult>['renderItem'] = ({ item }) => (
  <ResultItem item={item} />
)

const keyExtractor: FlatListProps<QuestionResult>['keyExtractor'] = ({ id }) =>
  id

const getItemLayout: FlatListProps<QuestionResult>['getItemLayout'] = (
  _,
  index,
) => ({
  index,
  length: ITEM_HEIGHT,
  offset: ITEM_HEIGHT * index,
})

export const Results: SceneProps<Routes.Results> = ({ navigation }) => {
  const {
    Theme: { colors },
  } = useTheme()

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    })
  }, [colors.WHITE, navigation])

  const dispatch = useAppDispatch()

  const game = useSelector(selectGameConfiguration)

  const onPressRestartGame = () => {
    dispatch(restartGameAction())
    navigation.reset({ index: 0, routes: [{ name: Routes.ConfigureGame }] })
  }

  const total = game.questionAmount.valueOf()
  const correctOnes = game.results.filter(item => item.correct).length

  return (
    <SceneContainer
      barStyle={'light-content'}
      style={[styles.container, { backgroundColor: colors.SECONDARY }]}>
      <View style={{ ...StyleSheet.absoluteFillObject }}>
        <Svg width={width} height={height} viewBox="0 0 375 812" fill="none">
          <Path
            d="M-382.5 612.5L-835 243.5V948.5H834V264.5L487 673L39.5 -76L-382.5 612.5Z"
            fill="#142664"
            fillOpacity="0.4"
          />
        </Svg>
      </View>
      <FlatList
        ListHeaderComponent={<Header total={total} correctOnes={correctOnes} />}
        data={game.results}
        renderItem={renderItem}
        keyExtractor={keyExtractor}
        getItemLayout={getItemLayout}
      />
      <StyledContainer style={styles.footer}>
        <StyledButton
          mode={ButtonMode.DARK}
          label={localization('restartGameLabel')}
          onPress={onPressRestartGame}
        />
      </StyledContainer>
    </SceneContainer>
  )
}
