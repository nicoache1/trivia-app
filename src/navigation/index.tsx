import { createNativeStackNavigator } from '@react-navigation/native-stack'
import React from 'react'

import { AnimatedSplash } from '../scenes/animatedSplash'
import { ConfigureGame } from '../scenes/configureGame'
import { Game } from '../scenes/game'
import { Results } from '../scenes/results'
import { Welcome } from '../scenes/welcome'
import { Routes } from './routes'
import { ParamList } from './types'

const Stack = createNativeStackNavigator<ParamList>()

export const AppNavigator: React.FC<{}> = ({}) => {
  return (
    <Stack.Navigator
      screenOptions={{
        contentStyle: {
          backgroundColor: 'white',
        },
      }}>
      <Stack.Screen name={Routes.AnimatedSplash} component={AnimatedSplash} />
      <Stack.Screen
        name={Routes.Welcome}
        component={Welcome}
        options={{ presentation: 'modal' }}
      />
      <Stack.Screen name={Routes.ConfigureGame} component={ConfigureGame} />
      <Stack.Screen name={Routes.Game} component={Game} />
      <Stack.Screen name={Routes.Results} component={Results} />
    </Stack.Navigator>
  )
}
