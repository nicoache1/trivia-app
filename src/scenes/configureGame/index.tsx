import ConfigureImage from 'assets/configure.png'
import { localization } from 'localization'
import React, { useLayoutEffect } from 'react'
import { Controller, useForm } from 'react-hook-form'
import { Dimensions, Image, View } from 'react-native'
import { Background } from 'src/components/Background'
import { SceneContainer } from 'src/components/SceneContainer'
import { StyledButton } from 'src/components/StyledButton'
import { ButtonMode } from 'src/components/StyledButton/ButtonMode'
import { StyledContainer } from 'src/components/StyledContainer'
import { StyledText } from 'src/components/StyledText'
import { Routes } from 'src/navigation/routes'
import { SceneProps } from 'src/navigation/types'
import { useAppDispatch } from 'src/store'
import { setGameAction } from 'src/store/storage/game'
import { useTheme } from 'src/styles/Theme'
import { AppTypography, ColorPalette } from 'src/styles/types'
import { Game, getDifficulties, getQuestionAmounts } from 'src/types/game'

import { renderInputControl, renderSegmentedControl } from './formRenders'
import { styles } from './styles'
import { ConfigureGameForm } from './types'

const { width, height } = Dimensions.get('window')

export const ConfigureGame: SceneProps<Routes.ConfigureGame> = ({
  navigation,
}) => {
  const {
    Theme: { colors },
  } = useTheme()

  const dispatch = useAppDispatch()

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    })
  }, [navigation])

  const { control, handleSubmit } = useForm<ConfigureGameForm>({
    defaultValues: {
      amount: 0,
      difficulty: 0,
      name: localization('defaultName'),
    },
  })

  const difficulties = getDifficulties()
  const questions = getQuestionAmounts()

  const onSubmit = (data: ConfigureGameForm) => {
    const newGame: Game = {
      difficulty: difficulties[data.difficulty],
      player: data.name,
      questionAmount: questions[data.amount],
      results: [],
    }
    dispatch(setGameAction(newGame))
    navigation.navigate(Routes.Game)
  }

  return (
    <View
      style={[styles.screenContainer, { backgroundColor: colors.SECONDARY }]}>
      <SceneContainer style={styles.scene} barStyle={'light-content'}>
        <View style={styles.headerContainer}>
          <Background />
          <Image
            source={ConfigureImage}
            style={{ height: height / 2.5, width: width * 0.9 }}
            resizeMode={'contain'}
          />
          <View style={styles.textContainer}>
            <StyledText
              color={ColorPalette.TERTIARY}
              typography={AppTypography.H1}>
              {localization('configureTitle1')}
            </StyledText>
            <StyledText
              color={ColorPalette.WHITE}
              typography={AppTypography.H1}>
              {localization('configureTitle2')}
            </StyledText>
          </View>
        </View>
        <View style={styles.formContainer}>
          <Controller
            control={control}
            rules={{
              required: true,
            }}
            render={renderInputControl}
            name={'name'}
          />
          <StyledContainer>
            <View style={styles.subtitle}>
              <StyledText
                color={ColorPalette.WHITE}
                typography={AppTypography.P1_BOLD}>
                {localization('difficultyLabel')}
              </StyledText>
            </View>
            <Controller
              control={control}
              rules={{
                required: true,
              }}
              render={renderSegmentedControl<'difficulty'>(
                difficulties,
                colors.WHITE,
              )}
              name={'difficulty'}
            />
          </StyledContainer>
          <StyledContainer>
            <View style={styles.subtitle}>
              <StyledText
                color={ColorPalette.WHITE}
                typography={AppTypography.P1_BOLD}>
                {localization('questionAmountLabel')}
              </StyledText>
            </View>
            <Controller
              control={control}
              rules={{
                required: true,
              }}
              render={renderSegmentedControl<'amount'>(
                questions.map(item => `${item}`),
                colors.WHITE,
              )}
              name={'amount'}
            />
          </StyledContainer>
        </View>
        <StyledContainer style={styles.buttonContainer}>
          <StyledButton
            mode={ButtonMode.DARK}
            label={localization('getStartedLabel')}
            onPress={handleSubmit(onSubmit)}
          />
        </StyledContainer>
      </SceneContainer>
    </View>
  )
}
