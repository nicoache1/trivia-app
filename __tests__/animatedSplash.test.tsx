import { render } from '@testing-library/react-native'
import React from 'react'
import { Routes } from 'src/navigation/routes'
import { AnimatedSplash } from 'src/scenes/animatedSplash'

import {
  createNavigationPropMock,
  createRoutePropMock,
} from '../__mocks__/navigationMocks'

test('Expect AnimatedSplash to be rendered correctly with default', async () => {
  const mockRoute = createRoutePropMock<Routes.AnimatedSplash>(
    Routes.AnimatedSplash,
  )

  const mockNavigation = createNavigationPropMock<Routes.AnimatedSplash>()

  const renderedSnapshot = render(
    <AnimatedSplash route={mockRoute} navigation={mockNavigation} />,
  ).toJSON()

  expect(renderedSnapshot).toMatchSnapshot()
})
