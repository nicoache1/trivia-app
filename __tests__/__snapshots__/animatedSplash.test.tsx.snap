// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect AnimatedSplash to be rendered correctly with default 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Image
    animatedStyle={
      Object {
        "value": Object {
          "height": 70,
          "opacity": 1,
          "zIndex": 100,
        },
      }
    }
    collapsable={false}
    onLoadEnd={[Function]}
    resizeMode="contain"
    source={1}
    style={
      Object {
        "height": 70,
        "opacity": 1,
        "zIndex": 100,
      }
    }
  />
  <View
    animatedStyle={
      Object {
        "value": Object {
          "backgroundColor": "rgb(108,92,231)",
          "transform": Array [
            Object {
              "translateY": 375,
            },
            Object {
              "rotate": "-25deg",
            },
            Object {
              "translateX": -1334,
            },
            Object {
              "translateY": -375,
            },
          ],
          "zIndex": 1,
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "backgroundColor": "rgb(108,92,231)",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 375,
          },
          Object {
            "rotate": "-25deg",
          },
          Object {
            "translateX": -1334,
          },
          Object {
            "translateY": -375,
          },
        ],
        "zIndex": 1,
      }
    }
  />
</View>
`;
